shenyu-plugin/shenyu-plugin-wasm-base/src/test/java/org/apache/shenyu/plugin/wasm/base/handler/AbstractWasmDiscoveryHandlerTest.java

```diff
@@ -0,0 +1,133 @@
+    public void setUp() {
+
+        this.discoverySyncData = mock(DiscoverySyncData.class);
+        this.testWasmPluginDiscoveryHandler = new TestWasmPluginDiscoveryHandler();
+        this.discoveryUpstreamDataHandler = testWasmPluginDiscoveryHandler;
+        when(discoverySyncData.getSelectorId()).thenReturn("SHENYU");
+    }
+
+    /**
+     * The handlerDiscoveryUpstreamData test.
+     */
+    @Test
+    public void handlerDiscoveryUpstreamDataTest() {
+        discoveryUpstreamDataHandler = mock(DiscoveryUpstreamDataHandler.class);
+        discoveryUpstreamDataHandler.handlerDiscoveryUpstreamData(discoverySyncData);
+        testWasmPluginDiscoveryHandler.handlerDiscoveryUpstreamData(discoverySyncData);
+        verify(discoveryUpstreamDataHandler).handlerDiscoveryUpstreamData(discoverySyncData);
+
+    }
+
+    /**
+     * The plugin name test.
+     */
+    @Test
+    public void pluginNameTest() {
+        assertEquals("SHENYU_TEST", discoveryUpstreamDataHandler.pluginName());
+        assertEquals("SHENYU_TEST", testWasmPluginDiscoveryHandler.pluginName());
+    }
+
+    static class TestWasmPluginDiscoveryHandler extends AbstractWasmDiscoveryHandler {
+
+        private static final Map<Long, String> RESULTS = new ConcurrentHashMap<>();
+
+        @Override
+        protected Map<String, Func> initWasmCallJavaFunc(final Store<Void> store) {
+            Map<String, Func> funcMap = new HashMap<>();
+            funcMap.put("get_args", WasmFunctions.wrap(store, WasmValType.I64, WasmValType.I64, WasmValType.I32, WasmValType.I32,
+                (argId, addr, len) -> {
+                    String config = "hello from java " + argId;
+                    LOG.info("java side->" + config);
+                    assertEquals("hello from java 0", config);
+                    ByteBuffer buf = super.getBuffer();
+                    for (int i = 0; i < len && i < config.length(); i++) {
+                        buf.put(addr.intValue() + i, (byte) config.charAt(i));
+                    }
+                    return Math.min(config.length(), len);
+                }));
+            funcMap.put("put_result", WasmFunctions.wrap(store, WasmValType.I64, WasmValType.I64, WasmValType.I32, WasmValType.I32,
+                (argId, addr, len) -> {
+                    ByteBuffer buf = super.getBuffer();
+                    byte[] bytes = new byte[len];
+                    for (int i = 0; i < len; i++) {
+                        bytes[i] = buf.get(addr.intValue() + i);
+                    }
+                    String result = new String(bytes, StandardCharsets.UTF_8);
+                    assertEquals("rust result", result);
+                    RESULTS.put(argId, result);
+                    LOG.info("java side->" + result);
+                    return 0;
+                }));
+            return funcMap;
+        }
+
+        @Override
+        public void handlerDiscoveryUpstreamData(final DiscoverySyncData discoverySyncData) {
+        }
```

Comment:
We call WASM function in `AbstractWasmDiscoveryHandler#handlerDiscoveryUpstreamData`, but you override it.

Reply:
So how should I change this function (`AbstractWasmDiscoveryHandler#handlerDiscoveryUpstreamData`) ?

Reply:
Remove `TestWasmPluginDiscoveryHandler#handlerDiscoveryUpstreamData`.

Reply:
Do you mean something like this?
```diff
diff --git a/shenyu-plugin/shenyu-plugin-wasm-base/src/test/java/org/apache/shenyu/plugin/wasm/base/handler/AbstractWasmDiscoveryHandlerTest.java b/shenyu-plugin/shenyu-plugin-wasm-base/src/test/java/org/apache/shenyu/plugin/wasm/base/handler/AbstractWasmDiscoveryHandlerTest.java
index f3305adeb..07622b107 100644
--- a/shenyu-plugin/shenyu-plugin-wasm-base/src/test/java/org/apache/shenyu/plugin/wasm/base/handler/AbstractWasmDiscoveryHandlerTest.java
+++ b/shenyu-plugin/shenyu-plugin-wasm-base/src/test/java/org/apache/shenyu/plugin/wasm/base/handler/AbstractWasmDiscoveryHandlerTest.java
@@ -116,10 +116,6 @@ class AbstractWasmDiscoveryHandlerTest {
             return funcMap;
         }
 
-        @Override
-        public void handlerDiscoveryUpstreamData(final DiscoverySyncData discoverySyncData) {
-        }
-
         @Override
         public String pluginName() {
             return "SHENYU_TEST";

```

Reply:
Do you mean something like this but it will give a null pointer exception?
```diff
diff --git a/shenyu-plugin/shenyu-plugin-wasm-base/src/test/java/org/apache/shenyu/plugin/wasm/base/handler/AbstractWasmDiscoveryHandlerTest.java b/shenyu-plugin/shenyu-plugin-wasm-base/src/test/java/org/apache/shenyu/plugin/wasm/base/handler/AbstractWasmDiscoveryHandlerTest.java
index f3305adeb..07622b107 100644
--- a/shenyu-plugin/shenyu-plugin-wasm-base/src/test/java/org/apache/shenyu/plugin/wasm/base/handler/AbstractWasmDiscoveryHandlerTest.java
+++ b/shenyu-plugin/shenyu-plugin-wasm-base/src/test/java/org/apache/shenyu/plugin/wasm/base/handler/AbstractWasmDiscoveryHandlerTest.java
@@ -116,10 +116,6 @@ class AbstractWasmDiscoveryHandlerTest {
             return funcMap;
         }
 
-        @Override
-        public void handlerDiscoveryUpstreamData(final DiscoverySyncData discoverySyncData) {
-        }
-
         @Override
         public String pluginName() {
             return "SHENYU_TEST";

``` 
unless you want me to remove the test itself?
```diff
diff --git a/shenyu-plugin/shenyu-plugin-wasm-base/src/test/java/org/apache/shenyu/plugin/wasm/base/handler/AbstractWasmDiscoveryHandlerTest.java b/shenyu-plugin/shenyu-plugin-wasm-base/src/test/java/org/apache/shenyu/plugin/wasm/base/handler/AbstractWasmDiscoveryHandlerTest.java
index f3305adeb..262f5ecf7 100644
--- a/shenyu-plugin/shenyu-plugin-wasm-base/src/test/java/org/apache/shenyu/plugin/wasm/base/handler/AbstractWasmDiscoveryHandlerTest.java
+++ b/shenyu-plugin/shenyu-plugin-wasm-base/src/test/java/org/apache/shenyu/plugin/wasm/base/handler/AbstractWasmDiscoveryHandlerTest.java
@@ -61,18 +61,6 @@ class AbstractWasmDiscoveryHandlerTest {
         when(discoverySyncData.getSelectorId()).thenReturn("SHENYU");
     }
 
-    /**
-     * The handlerDiscoveryUpstreamData test.
-     */
-    @Test
-    public void handlerDiscoveryUpstreamDataTest() {
-        discoveryUpstreamDataHandler = mock(DiscoveryUpstreamDataHandler.class);
-        discoveryUpstreamDataHandler.handlerDiscoveryUpstreamData(discoverySyncData);
-        testWasmPluginDiscoveryHandler.handlerDiscoveryUpstreamData(discoverySyncData);
-        verify(discoveryUpstreamDataHandler).handlerDiscoveryUpstreamData(discoverySyncData);
-
-    }
-
     /**
      * The plugin name test.
      */
@@ -116,9 +104,7 @@ class AbstractWasmDiscoveryHandlerTest {
             return funcMap;
         }
 
-        @Override
-        public void handlerDiscoveryUpstreamData(final DiscoverySyncData discoverySyncData) {
-        }
+
 
         @Override
         public String pluginName() {

``` 
Do you want the `TestWasmPluginDiscoveryHandler`  class to look something like this?
```java
static class TestWasmPluginDiscoveryHandler extends AbstractWasmDiscoveryHandler {

        private static final Map<Long, String> RESULTS = new ConcurrentHashMap<>();

        @Override
        protected Map<String, Func> initWasmCallJavaFunc(final Store<Void> store) {
            Map<String, Func> funcMap = new HashMap<>();
            funcMap.put("get_args", WasmFunctions.wrap(store, WasmValType.I64, WasmValType.I64, WasmValType.I32, WasmValType.I32,
                (argId, addr, len) -> {
                    String config = "hello from java " + argId;
                    LOG.info("java side->" + config);
                    assertEquals("hello from java 0", config);
                    ByteBuffer buf = super.getBuffer();
                    for (int i = 0; i < len && i < config.length(); i++) {
                        buf.put(addr.intValue() + i, (byte) config.charAt(i));
                    }
                    return Math.min(config.length(), len);
                }));
            funcMap.put("put_result", WasmFunctions.wrap(store, WasmValType.I64, WasmValType.I64, WasmValType.I32, WasmValType.I32,
                (argId, addr, len) -> {
                    ByteBuffer buf = super.getBuffer();
                    byte[] bytes = new byte[len];
                    for (int i = 0; i < len; i++) {
                        bytes[i] = buf.get(addr.intValue() + i);
                    }
                    String result = new String(bytes, StandardCharsets.UTF_8);
                    assertEquals("rust result", result);
                    RESULTS.put(argId, result);
                    LOG.info("java side->" + result);
                    return 0;
                }));
            return funcMap;
        }

        @Override
        public String pluginName() {
            return "SHENYU_TEST";
        }

        @Override
        protected Long getArgumentId(final DiscoverySyncData discoverySyncData) {
            return null;
        }
    }
}

``` 

This way.